File: java/linkedlist/DetectCycleStart.java
/**
 * Q11. Linked List Cycle II
 * 
 * Problem:
 * Given the head of a linked list, return the node where the cycle begins.
 * If there is no cycle, return null.
 * 
 * There is a cycle in a linked list if there is some node in the list that can be reached again
 * by continuously following the next pointer. Internally, pos is used to denote the index of the
 * node that tail's next pointer is connected to (0-indexed). It is -1 if there is no cycle.
 * Note that pos is not passed as a parameter.
 * 
 * Example:
 * Input: head = [3,2,0,-4], pos = 1
 * Output: tail connects to node index 1
 * Explanation: There is a cycle in the linked list, where tail connects to the second node.
 * 
 * Approach:
 * 1. Use Floyd's Cycle Detection Algorithm (Tortoise and Hare)
 * 2. When fast and slow meet, reset slow to head
 * 3. Move both pointers one step at a time
 * 4. They will meet at the cycle start
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use two pointers
 */

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class DetectCycleStart {
    public ListNode detectCycle(ListNode head) {
        // Handle edge cases
        if (head == null || head.next == null) {
            return null;
        }

        // Step 1: Find meeting point using Floyd's algorithm
        ListNode fast = head;
        ListNode slow = head;
        
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            
            // If they meet, there is a cycle
            if (fast == slow) {
                // Step 2: Find cycle start
                ListNode current = head;
                while (current != slow) {
                    current = current.next;
                    slow = slow.next;
                }
                return current;
            }
        }
        
        // No cycle found
        return null;
    }

    // Main method for testing
    public static void main(String[] args) {
        // Test case 1: List with cycle at index 1
        ListNode head1 = new ListNode(3);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(0);
        ListNode node4 = new ListNode(-4);
        
        head1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node2; // Create cycle at node2

        DetectCycleStart solution = new DetectCycleStart();
        ListNode cycleStart = solution.detectCycle(head1);
        System.out.println("Cycle starts at node with value: " + 
            (cycleStart != null ? cycleStart.val : "null"));

        // Test case 2: List with cycle at head
        ListNode head2 = new ListNode(1);
        ListNode node5 = new ListNode(2);
        head2.next = node5;
        node5.next = head2; // Create cycle at head

        cycleStart = solution.detectCycle(head2);
        System.out.println("Cycle starts at node with value: " + 
            (cycleStart != null ? cycleStart.val : "null"));

        // Test case 3: List without cycle
        ListNode head3 = new ListNode(1);
        head3.next = new ListNode(2);
        head3.next.next = new ListNode(3);

        cycleStart = solution.detectCycle(head3);
        System.out.println("Cycle starts at node with value: " + 
            (cycleStart != null ? cycleStart.val : "null"));

        // Test case 4: Empty list
        cycleStart = solution.detectCycle(null);
        System.out.println("Cycle starts at node with value: " + 
            (cycleStart != null ? cycleStart.val : "null"));

        // Test case 5: Single node without cycle
        ListNode head5 = new ListNode(1);
        cycleStart = solution.detectCycle(head5);
        System.out.println("Cycle starts at node with value: " + 
            (cycleStart != null ? cycleStart.val : "null"));
    }
} 

File: javascript/linkedlist/DetectCycleStart.js
/**
 * Detect Cycle Start
 * 
 * Problem:
 * Given the head of a linked list, return the node where the cycle begins.
 * If there is no cycle, return null.
 * 
 * Example:
 * Input: head = [3,2,0,-4] (where -4 points back to 2)
 * Output: Node with value 2
 * Explanation:
 * - There is a cycle in the linked list
 * - The cycle starts at node with value 2
 * - The tail connects to the second node
 * 
 * Approach:
 * 1. Use Floyd's Cycle Detection Algorithm to find if there's a cycle
 * 2. If a cycle is found, use two pointers:
 *    - One at the meeting point
 *    - One at the head
 * 3. Move both pointers one step at a time until they meet
 * 4. The meeting point is the start of the cycle
 * 
 * Time Complexity: O(N) where N is the number of nodes
 * Space Complexity: O(1) as we only use pointers
 */

class ListNode {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

class DetectCycleStart {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    detectCycleStart(head) {
        if (!head || !head.next) return null;

        // Find meeting point using Floyd's Cycle Detection
        let slow = head;
        let fast = head;
        let hasCycle = false;

        while (fast && fast.next) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow === fast) {
                hasCycle = true;
                break;
            }
        }

        if (!hasCycle) return null;

        // Find the start of the cycle
        slow = head;
        while (slow !== fast) {
            slow = slow.next;
            fast = fast.next;
        }

        return slow;
    }

    // Helper method to create a linked list with a cycle
    static createListWithCycle(values, cycleStartIndex) {
        if (!values || values.length === 0) return null;
        
        const head = new ListNode(values[0]);
        let current = head;
        let cycleStart = null;
        
        for (let i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            if (i === cycleStartIndex) {
                cycleStart = current.next;
            }
            current = current.next;
        }
        
        // Create cycle if cycleStartIndex is valid
        if (cycleStartIndex >= 0 && cycleStartIndex < values.length) {
            current.next = cycleStart;
        }
        
        return head;
    }

    // Helper method to print the list (only works for lists without cycles)
    static printList(head) {
        let current = head;
        const values = [];
        const visited = new Set();
        
        while (current && !visited.has(current)) {
            values.push(current.val);
            visited.add(current);
            current = current.next;
        }
        
        if (current) {
            values.push('... (cycle)');
        }
        
        console.log(values.join(' -> '));
    }
}

// Test cases
function runTests() {
    const solution = new DetectCycleStart();

    // Test case 1: Regular case with cycle
    console.log("Test case 1: Regular case with cycle");
    const list1 = DetectCycleStart.createListWithCycle([3, 2, 0, -4], 1);
    console.log("List (with cycle at index 1):");
    DetectCycleStart.printList(list1);
    const cycleStart1 = solution.detectCycleStart(list1);
    console.log("Cycle starts at node with value:", cycleStart1 ? cycleStart1.val : null);

    // Test case 2: List without cycle
    console.log("\nTest case 2: List without cycle");
    const list2 = DetectCycleStart.createListWithCycle([1, 2, 3, 4], -1);
    console.log("List (no cycle):");
    DetectCycleStart.printList(list2);
    const cycleStart2 = solution.detectCycleStart(list2);
    console.log("Cycle starts at node with value:", cycleStart2 ? cycleStart2.val : null);

    // Test case 3: Single node with self-loop
    console.log("\nTest case 3: Single node with self-loop");
    const list3 = DetectCycleStart.createListWithCycle([1], 0);
    console.log("List (self-loop):");
    DetectCycleStart.printList(list3);
    const cycleStart3 = solution.detectCycleStart(list3);
    console.log("Cycle starts at node with value:", cycleStart3 ? cycleStart3.val : null);

    // Test case 4: Empty list
    console.log("\nTest case 4: Empty list");
    const list4 = null;
    console.log("List: null");
    const cycleStart4 = solution.detectCycleStart(list4);
    console.log("Cycle starts at node with value:", cycleStart4 ? cycleStart4.val : null);

    // Test case 5: Two node list with cycle
    console.log("\nTest case 5: Two node list with cycle");
    const list5 = DetectCycleStart.createListWithCycle([1, 2], 0);
    console.log("List (cycle at start):");
    DetectCycleStart.printList(list5);
    const cycleStart5 = solution.detectCycleStart(list5);
    console.log("Cycle starts at node with value:", cycleStart5 ? cycleStart5.val : null);

    // Test case 6: Long list with cycle
    console.log("\nTest case 6: Long list with cycle");
    const list6 = DetectCycleStart.createListWithCycle([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5);
    console.log("List (with cycle at index 5):");
    DetectCycleStart.printList(list6);
    const cycleStart6 = solution.detectCycleStart(list6);
    console.log("Cycle starts at node with value:", cycleStart6 ? cycleStart6.val : null);
}

// Run the tests
runTests(); 